gobblin-runtime/src/main/java/gobblin/runtime/locks/ZookeeperBasedJobLock.java

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright (C) 2014-2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.runtime.locks;
+
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.CuratorFrameworkFactory;
+import org.apache.curator.framework.recipes.locks.InterProcessSemaphoreMutex;
+import org.apache.curator.framework.state.ConnectionState;
+import org.apache.curator.framework.state.ConnectionStateListener;
+import org.apache.curator.retry.ExponentialBackoffRetry;
+
+import com.google.common.collect.Maps;
+
+import gobblin.configuration.ConfigurationKeys;
+
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * An implementation of {@link JobLock} that uses Zookeeper.
+ *
+ * @author Joel Baranick
+ */
+@Slf4j
+public class ZookeeperBasedJobLock implements JobLock {
+  private static final String LOCKS_ROOT_PATH = "/locks";
+  private static final String CONNECTION_STRING_DEFAULT = "localhost:2181";
+  private static final int LOCKS_ACQUIRE_TIMEOUT_MILLISECONDS_DEFAULT = 5000;
+  private static final int CONNECTION_TIMEOUT_SECONDS_DEFAULT = 30;
+  private static final int SESSION_TIMEOUT_SECONDS_DEFAULT = 180;
+  private static final int RETRY_BACKOFF_SECONDS_DEFAULT = 1;
+  private static final int MAX_RETRY_COUNT_DEFAULT = 10;
+  private static CuratorFramework curatorFramework;
+  private static ConcurrentMap<String, JobLockEventListener> lockEventListeners = Maps.newConcurrentMap();
+  private static Thread curatorFrameworkShutdownHook;
+
+  public static final String LOCKS_ACQUIRE_TIMEOUT_MILLISECONDS = "zookeeper.locks.acquire.timeout.milliseconds";
+  public static final String CONNECTION_STRING = "zookeeper.connection.string";
+  public static final String CONNECTION_TIMEOUT_SECONDS = "zookeeper.connection.timeout.seconds";
+  public static final String SESSION_TIMEOUT_SECONDS = "zookeeper.session.timeout.seconds";
+  public static final String RETRY_BACKOFF_SECONDS = "zookeeper.retry.backoff.seconds";
+  public static final String MAX_RETRY_COUNT = "zookeeper.retry.count.max";
```

Comment:
We are trying to make any new configuration keys compatible with Typesafe Config - that being said dots should no longer be substitutions for spaces, when defining a new config key. So this configs should read something like `gobblin.locks.zookeeper.retry.max_count` and `gobblin.locks.zookeeper.retry.backoff_seconds`